python/ray/serve/backend_state.py

```diff
@@ -596,22 +621,35 @@ def delete_backend(self, backend_tag: BackendTag,
         return new_goal_id
 
     def _stop_wrong_version_replicas(
-            self, backend_tag: BackendTag, version: str,
-            graceful_shutdown_timeout_s: float) -> int:
+            self, backend_tag: BackendTag, target_replicas: int,
+            target_version: str, graceful_shutdown_timeout_s: float) -> int:
         # NOTE(edoakes): this short-circuits when using the legacy
         # `create_backend` codepath -- it can be removed once we deprecate
         # that as the version should never be None.
-        if version is None:
+        if target_version is None:
             return
 
-        # TODO(edoakes): to implement rolling upgrade, all we should need to
-        # do is cap the number of old version replicas that are stopped here.
-        replicas_to_stop = self._replicas[backend_tag].pop(
-            exclude_version=version,
+        old_running_replicas = self._replicas[backend_tag].count(
+            exclude_version=target_version,
             states=[
                 ReplicaState.SHOULD_START, ReplicaState.STARTING,
                 ReplicaState.RUNNING
             ])
+        new_running_replicas = self._replicas[backend_tag].count(
+            version=target_version, states=[ReplicaState.RUNNING])
+
+        pending_replicas = (
```

Comment:
we can assert that this is equal to .count(version=target_version, states=[SHOULD_START,STARTING])?

Reply:
It might be more manageable if we separate the rolling upgrade logic from the scaling logic instead of doing them both in _scale_backend_replicas

Reply:
So the problem here is that once replicas are completely stopped, they are removed from the state data structure, which means you can't use the equality check that @architkulkarni suggested (you could make it a <= though). That's why I'm "reverse engineering" the pending replicas instead of directly counting the number stopping and number starting.

Good point about the negative pending_replicas @architkulkarni. I agree that it would be better to separate this logic out, let me see if there's a clean way to do that. It'd also help with unit testing for both methods.